storage/reads/merge.go

```diff
@@ -122,42 +122,81 @@ func (h *resultSetHeap) Pop() interface{} {
 	return item
 }
 
+// MergedStringIterator merges multiple storage.StringIterators into one.
+// It sorts and deduplicates adjacent values, so the output is sorted iff all inputs are sorted.
+// If all inputs are not sorted, then output order and deduplication are undefined and unpleasant.
 type MergedStringIterator struct {
-	iterators    []storage.StringIterator
-	uniqueValues map[string]struct{}
-	nextValue    string
+	heap      *stringIteratorHeap
+	nextValue string
 }
 
 func NewMergedStringIterator(iterators []storage.StringIterator) *MergedStringIterator {
+	nonEmptyIterators := make([]storage.StringIterator, 0, len(iterators))
+
+	for _, iterator := range iterators {
+		// All iterators must be Next()'d so that their Value() methods return a meaningful value, and sort properly.
+		if iterator.Next() {
+			nonEmptyIterators = append(nonEmptyIterators, iterator)
+		}
+	}
+	h := &stringIteratorHeap{iterators: nonEmptyIterators}
+	heap.Init(h)
+
 	return &MergedStringIterator{
-		iterators:    iterators,
-		uniqueValues: make(map[string]struct{}),
+		heap: h,
 	}
 }
 
-func (mr *MergedStringIterator) Next() bool {
-	// TODO assume that each iterator is sorted, and iterate in sorted order
-	// https://github.com/influxdata/influxdb/issues/13440
-	for len(mr.iterators) > 0 {
-		iterator := mr.iterators[0]
+func (msi *MergedStringIterator) Next() bool {
+	for msi.heap.Len() > 0 {
+		iterator := heap.Pop(msi.heap).(storage.StringIterator)
```

Comment:
Not a required change, however an alternative approach, using `heap.Fix`. 

A property of the min-heap is that element zero is always the minimum, so you can avoid two heap orderings per iteration (`heap.Pop`, `heap.Push`):

```go
for msi.heap.Len() > 0 {
	iterator := msi.heap.iterators[0]

	haveNext := false
	if proposedNextValue := iterator.Value(); proposedNextValue != msi.nextValue { // Skip dupes.
		msi.nextValue = proposedNextValue
		haveNext = true
	}

	if iterator.Next() {
		heap.Fix(msi.heap, 0)
	} else {
		heap.Pop(msi.heap)
	}

	if haveNext {
		return true
	}
}

return false
```